### PR TITLE
Call ProduceEnd() before consuming request body if response has already started (#1530)

### PR DESCRIPTION
**Pull request by cesarbs**
#1530

Prevents hanging clients waiting for the chunk terminator while server consumes the rest of the request body.

See https://github.com/dotnet/corefx/issues/17330#issuecomment-288248663.

cc @stephentoub

### CODE REVIEW

**Comment by davidfowl** [comment]
@cesarbs can you rebase on top of @halter73 's change

```
Failed   Microsoft.AspNetCore.Server.Kestrel.FunctionalTests.AddressRegistrationTests.RegisterAddresses_IPv6ScopeId_Success(addressInput: "http://[fe80::1208:2b32:1916:e4b0%utun0]:0/", testUrls: Func`2 { Method = System.String[] GetTestUrls(Microsoft.AspNetCore.Hosting.Server.Features.IServerAddressesFeature), Target = null })
```

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/FrameOfT.cs [hunk 1]
```diff
@@ -133,16 +133,28 @@ public Frame(IHttpApplication<TContext> application, ConnectionContext context)
                             {
                                 ResumeStreams();
 
+                                if (HasResponseStarted)
+                                {
+                                    // If the response has already started, call ProduceEnd() before
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/FrameOfT.cs]: Point to the bug in here

**Comment by cesarbs** [comment]
@davidfowl Rebased.

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/FrameOfT.cs (hunk 1 above)
Comment (cesarbs) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/FrameOfT.cs]: Pointed to the explanation in the corefx bug.

File: test/Microsoft.AspNetCore.Server.Kestrel.FunctionalTests/ResponseTests.cs [hunk 2]
```diff
@@ -1368,6 +1368,116 @@ public Task ResponseStatusCodeSetBeforeHttpContextDisposedRequestMalformedReadIg
...
 
+        [Fact]
+        public async Task WhenResponseAlreadyStartedResponseEndedBeforeDrainingRequestBody()
+        {
+            using (var server = new TestServer(async httpContext =>
+            {
+                await httpContext.Response.WriteAsync("hello, world");
+
+                // Force response start
+                await httpContext.Response.Body.FlushAsync();
```
Comment (davidfowl) [inline test/Microsoft.AspNetCore.Server.Kestrel.FunctionalTests/ResponseTests.cs]: You don't have to call FlushAsync to start the response, that happens on write

**Comment by davidfowl** [comment]
Can you make sure we have a test for all of the relevant message body types? (Opaque, ContentLength, Chunked)

File: test/Microsoft.AspNetCore.Server.Kestrel.FunctionalTests/ResponseTests.cs [hunk 3]
```diff
@@ -1368,6 +1368,113 @@ public Task ResponseStatusCodeSetBeforeHttpContextDisposedRequestMalformedReadIg
...
+                        $"Transfer-Encoding: chunked",
+                        "",
+                        "c",
+                        "hello, world",
+                        "");
+
+                    // If the expected behavior is regressed, this will hang because the
+                    // server will try to consume the request body before flushing the chunked
+                    // terminator.
+                    await connection.Receive(
```
Comment (halter73) [inline test/Microsoft.AspNetCore.Server.Kestrel.FunctionalTests/ResponseTests.cs]: Did you split this into two calls to connection.Receive() just to insert the above comment?

File: test/Microsoft.AspNetCore.Server.Kestrel.FunctionalTests/ResponseTests.cs [hunk 4]
```diff
@@ -1368,6 +1368,113 @@ public Task ResponseStatusCodeSetBeforeHttpContextDisposedRequestMalformedReadIg
...
+                        "0",
+                        "",
+                        "");
+                }
+            }
+        }
+
+        [Fact]
+        public async Task WhenResponseNotStartedResponseEndedAfterDrainingRequestBody()
```
Comment (halter73) [inline test/Microsoft.AspNetCore.Server.Kestrel.FunctionalTests/ResponseTests.cs]: Good test 👍

File: test/Microsoft.AspNetCore.Server.Kestrel.FunctionalTests/ResponseTests.cs [hunk 5]
```diff
@@ -1368,6 +1368,113 @@ public Task ResponseStatusCodeSetBeforeHttpContextDisposedRequestMalformedReadIg
...
+                        $"Date: {server.Context.DateHeaderValue}",
+                        "Content-Length: 0",
+                        "",
+                        "");
+                }
+            }
+        }
+
+        [Fact]
+        public async Task Sending100ContinueDoesNotStartResponse()
```
Comment (halter73) [inline test/Microsoft.AspNetCore.Server.Kestrel.FunctionalTests/ResponseTests.cs]: Can we test sending the chunk terminator pre-consume when we send a 100-continue and the app sends a response?

File: test/Microsoft.AspNetCore.Server.Kestrel.FunctionalTests/ResponseTests.cs (hunk 3 above)
Comment (cesarbs) [inline test/Microsoft.AspNetCore.Server.Kestrel.FunctionalTests/ResponseTests.cs]: No. It's to make it clear the first part will be received fine, but if there's a regression the connection will time out waiting for the chunk terminator.

File: test/Microsoft.AspNetCore.Server.Kestrel.FunctionalTests/ResponseTests.cs (hunk 5 above)
Comment (cesarbs) [inline test/Microsoft.AspNetCore.Server.Kestrel.FunctionalTests/ResponseTests.cs]: Added.

**Comment by cesarbs** [comment]
Updated.

**Comment by cesarbs** [comment]
@davidfowl This is unrelated to the message body type, so having tests per type won't add value. What we care about here is the order in which things happen.

**Comment by cesarbs** [comment]
📢

**Pull request merged**